src/MongoDB.AspNetCore.OData/MongoEnableQueryAttribute.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright 2023-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.OData.Extensions;
+using Microsoft.AspNetCore.OData.Query;
+using Microsoft.OData.Edm;
+using MongoDB.Driver;
+using MongoDB.Driver.Linq;
+
+namespace MongoDB.AspNetCore.OData;
+
+public sealed class MongoEnableQueryAttribute : EnableQueryAttribute
+{
+    private static readonly MethodInfo s_applyQueryInternalMethodInfo = typeof(MongoEnableQueryAttribute).GetMethod(
+        nameof(ApplyQueryInternal),
+        BindingFlags.Static | BindingFlags.NonPublic);
+    private static readonly MethodInfo s_applyProjectionMethodInfo = typeof(MongoEnableQueryAttribute).GetMethod(
+        nameof(ApplyProjection),
+        BindingFlags.Static | BindingFlags.NonPublic);
+
+    public MongoEnableQueryAttribute()
+    {
+        AllowedQueryOptions =
+            AllowedQueryOptions.Filter |
+            AllowedQueryOptions.Expand |
+            AllowedQueryOptions.Select |
+            AllowedQueryOptions.OrderBy |
+            AllowedQueryOptions.Top |
+            AllowedQueryOptions.Skip |
+            AllowedQueryOptions.Count |
+            AllowedQueryOptions.Compute;
+    }
+
+    public override void ValidateQuery(HttpRequest request, ODataQueryOptions queryOptions)
+    {
+        base.ValidateQuery(request, queryOptions);
+
+        // Store this into the http context in case we will need to enforce projection
+        request.HttpContext.Items[nameof(ODataQueryOptions)] = queryOptions;
+    }
+
+    public override void OnActionExecuted(ActionExecutedContext actionExecutedContext)
+    {
+        base.OnActionExecuted(actionExecutedContext);
+
+        var response = actionExecutedContext.HttpContext.Response;
+        // Check whether the response is set and successful
+        if (response != null && response.IsSuccessStatusCode() && actionExecutedContext.Result is ObjectResult responseContent)
+        {
+            // If response is IMongoQueryable it means default implementation bypasses all transformations,
+            // we want force apply projection to limit MQL request to struct properties only
```

Comment:
Should this be: "to structural properties"?
Not to be confused with structs.

Reply:
Done